evm/src/executor/inspector/tracer.rs

```diff
@@ -79,8 +79,17 @@ impl Tracer {
     }
 
     pub fn fill_step(&mut self, gas: u64, status: Return) {
-        let trace = &mut self.traces.arena
-            [*self.trace_stack.last().expect("can't fill step without starting a trace first")];
+        let last_trace_idx =
+            *self.trace_stack.last().expect("can't fill step without starting a trace first");
+        let trace_idx = if !self.traces.arena[last_trace_idx].trace.step_stack.is_empty() {
+            last_trace_idx
+        } else {
+            // Special case when we started new trace between start_step() and fill_step(), so we
+            // need to take the penultimate trace from the stack instead of the last
+            self.trace_stack[self.trace_stack.len() - 2]
```

Comment:
special cases -.-

Let's go with that but maybe use `get().expect()` so we get a better error if that doesn't hold, or is this always safe?

Reply:
ok there was a better way indeed :D fixed